### PR TITLE
Update snowflake algorithm to support modulus of 2^n

### PR DESCRIPTION
**Pull request by geomonlin**
Fixes ##3422.

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java [hunk 1]
```diff
@@ -78,7 +80,7 @@
     @Setter
     private Properties properties = new Properties();
     
-    private byte sequenceOffset;
+    private long sequenceOffset;
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java]: it is unnecessary for long with sequenceOffset, int is enough

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java [hunk 2]
```diff
@@ -68,6 +68,8 @@
     private static final long WORKER_ID_MAX_VALUE = 1L << WORKER_ID_BITS;
     
     private static final long WORKER_ID = 0;
+
+    private static final long VIBRATION_VALUE = 1;
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java]: should be DEFAULT_VIBRATION_VALUE?

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java [hunk 3]
```diff
@@ -134,7 +136,13 @@ private long getWorkerId() {
         Preconditions.checkArgument(result >= 0L && result < WORKER_ID_MAX_VALUE);
         return result;
     }
-    
+
+    private long getCustomVibrationValue() {
+        long result = Long.valueOf(properties.getProperty("custom.vibration.value", String.valueOf(VIBRATION_VALUE)));
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java]: should need change other name of "custom.vibration.value", maybe "max.vibration.offset" is better

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java [hunk 4]
```diff
@@ -134,7 +136,13 @@ private long getWorkerId() {
...
+        Preconditions.checkArgument(result > 0L && result <= SEQUENCE_MASK,"Illegal custom vibration value");
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java]: checkstyle error for comma without blank

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java [hunk 5]
```diff
@@ -148,6 +156,15 @@ private long waitUntilNextTime(final long lastTime) {
     }
     
     private void vibrateSequenceOffset() {
-        sequenceOffset = (byte) (~sequenceOffset & 1);
+        sequenceOffset = (sequenceOffset + 1) & getCustomVibrationValue();
+    }
+
+    private long optimizationCustomVibrationValue(long vibrationValue) {
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java]: This method is hard to understander too, what is the proposal?

File: sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java [hunk 6]
```diff
@@ -148,6 +156,15 @@ private long waitUntilNextTime(final long lastTime) {
     }
     
     private void vibrateSequenceOffset() {
-        sequenceOffset = (byte) (~sequenceOffset & 1);
+        sequenceOffset = (sequenceOffset + 1) & getCustomVibrationValue();
```
Comment (terrymanu) [inline sharding-core/sharding-core-common/src/main/java/org/apache/shardingsphere/core/strategy/keygen/SnowflakeShardingKeyGenerator.java]: This method is hard to understander, what is the proposal?

**Comment by terrymanu** [comment]
Why change so many files?

**Comment by geomonlin** [comment]
Because the historical commit record was committed carelessly when the branch was updated. I close the request for resubmission. 1623537355 邮箱1623537355@qq.com 签名由 网易邮箱大师 定制 On 11/09/2019 00:31, Liang Zhang wrote: Why change so many files? — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe.

**Pull request closed without merging**